bitwarden/server
# Use sas token for attachment downloads

# Overview

This PR add a `GetAttachmentDownloadUrlAsync` method, which clients will need to use to receive a valid, time-limited download URL directly from Azure.

## Backwards compatibility

These changes are backwards compatible with clients not requesting attachments made **before** these changes are in place.

However, once these changes are in place, new attachments will be created in a blobs which are not publicly accessible. The default URLs created will no longer work. This means that **new attachments will not be backward compatible with old clients** which do not use the new `GetAttachmentDownloadUrlAsync` method to receive a temporary download URL.

Additionally, any attachment already created will receive a time-limited URL, but they are still publicly accessible blobs, so they are not _really_ locked to a 1 min expiration.

# TODO

The following will be completed as separate PRs

* Add similar method to Send
* Add direct attachment upload URL
* Add direct Send upload URL
* PRs to all clients to use `GetAttachmentDownloadUrlAsync` and similar methods established above

# Files Changed

There are a huge number of changes due to moving GlobalSettings to a subdirectory, new namespace, and implementing an interface. These are benign additions of a using statement that I'll ignore below.

* **CiphersController.cs**: Add GetAttachmentData API endpoint which, primarily provides a time-limited Access URL.
* **CipherAttachment.cs**: Add ContainerName and AttachmentId to Attachment MetaData. The ContainerName defaults to the old `attachments` container if not present. AttachmentID is not serialized, but is populated in `Table/Cipher.cs` after deserialization to provide a one-stop-shop for attachment metadata
* **Table/Cipher.cs**: Use `DefaultValue` attribute when deserializing `CipherAttachment.MetaData`. Also fills in `AttachmentId` for ease-of-use.
* **IAttachmentStorageService**: Use CipherAttachment.MetaData to also pass container name for AzureStorageService to provide the correct blob. Add `GetAttachmentDownloadUrlAsync` method to provide temporary download links.
* **AzureAttachmentStorageService**: Allow for multiple attachment containers and initialize required containers. Delete methods now delete from all known containers. Use SAS to provide a time-limited URL to access attachment.
* **CipherService**: Generally provide container name to AttachmentStorageService methods
  * Upgrade methods update Attachment metadata to allow for potential changes to metadata
  * Save Attachments provide MetaData to allow for alteration by AttachmentStorageService prior to saving the Cipher
* **LocalAttachmentStorageService**: Centralize methods for path formation, use Path.Combine instead of serialized strings. Implement `GetAttachmentDownloadUrlAsync`.
* **NoopAttachmentStorageService**: Add `GetAttachmentDownloadUrlAsync`
* **GlobalSettings/IGlobalSettings/IFileStorageSettings**: Use Interface for settings to be tested. These interfaces can be expanded upon when needed for future testing.
* **ServiceCollectionExtension.cs**: Add IGlobalSettings Singleton dependency which provides already existing GlobalSettings Singleton.

## Tests
* **TempDirector**: Disposable temp directory that cleans itself up upon Dispose
* **LocalAttachmentStorageService**: Test equivalency of changes
* **CipherAttachmentMetaDataFixtures**: Add AutoFixture customizations for providing Attachments in different states.
  * New style attachments
  * Old attachments, which do not specify a ContainerName
  * Old attachments, which do not specify a ContainerName, but have been deserialized to the default value.
  * Very old attachments, which do not have a key associated with them

# Test Concerns

I have not yet been able to Unit Test AzureAttachmentStorageService. This is because the Azure methods do not return interfaces, so they're hard to test without actual Azure access.

We'll need manual testing around:
* Sharing old attachments which don't have a key (this might requires db manipulation, these attachments are probably pretty rare)
* Sharing attachments which have a key
* Creating attachments
* Deleting attachments (verify deleted in Azure)
* Downloading newly created attachments -- remove SAS query key, fails
* Downloading any attachment -- wait > 1 min to download
* Use old client to download an old attachment
* Use new client to download an old attachment
